src/linalg.cpp

```diff
@@ -0,0 +1,319 @@
+
+                /* Get the column index */
+                col = JA[i];
+
+                if (row == col)
+                  x[row] += SOR_factor * b[row] / DIAG[row];
+                else
+                  x[row] -= SOR_factor * a[i] * x[col] / DIAG[row];
+              }
+            }
+          }
+        }
+      }
+    }
+
+    /* Compute the new source */
+    matrixMultiplication(M, X, &new_source);
+
+    /* Compute the residual */
+    residual = computeRMSE(&new_source, &old_source, true);
+
+    /* Copy the new source to the old source */
+    new_source.copyTo(&old_source);
+
+    /* Increment the interations counter */
+    iter++;
+
+    log_printf(INFO, "SOR iter: %i, res: %g", iter, residual);
+
+    if (residual < tol && iter > 10)
+      break;
+  }
+
+  log_printf(INFO, "linear solve iterations: %i", iter);
+}
+
+
+/**
+ * @brief Performs a matrix vector multiplication.
+ * @details This function takes in a Matrix (A), a variable Vector (X),
+ *          and a solution Vector (B) and computes the matrix vector product.
+ *          The solution Vector is modified in place.
+ * @param A a Matrix object
+ * @param X the variable Vector object
+ * @param B the solution Vector object
+ */
+void matrixMultiplication(Matrix* A, Vector* X, Vector* B) {
+
+  /* Check for consistency of matrix and vector dimensions */
+  if (A->getNumX() != B->getNumX() || A->getNumX() != X->getNumX())
+    log_printf(ERROR, "Cannot perform matrix multiplication  with different x "
+               "dimensions for the A matrix, B vector, and X vector: "
+               "(%i, %i, %i)", A->getNumX(), B->getNumX(), X->getNumX());
+  else if (A->getNumX() != B->getNumY() || A->getNumY() != X->getNumY())
+    log_printf(ERROR, "Cannot perform matrix multiplication with different y "
+               "dimensions for the A matrix, B vector, and X vector: "
+               "(%i, %i, %i)", A->getNumY(), B->getNumY(), X->getNumY());
+  else if (A->getNumGroups() != B->getNumGroups() ||
+           A->getNumGroups() != X->getNumGroups())
+    log_printf(ERROR, "Cannot perform matrix multiplication with different "
+               "num groups for the A matrix, M matrix, and X vector: "
+               "(%i, %i, %i)", A->getNumGroups(), B->getNumGroups(),
+               X->getNumGroups());
+
+  B->setAll(0.0);
+  int* IA = A->getIA();
+  int* JA = A->getJA();
+  FP_PRECISION* a = A->getA();
+  FP_PRECISION* x = X->getArray();
+  FP_PRECISION* b = B->getArray();
+  int num_rows = X->getNumRows();
+
+  #pragma omp parallel for
+  for (int row = 0; row < num_rows; row++) {
+    for (int i = IA[row]; i < IA[row+1]; i++)
+      b[row] += a[i] * x[JA[i]];
+  }
+}
+
+
+/**
+ * @brief Computes the Root Mean Square Error of two Vectors.
+ * @details This function takes in two vectors (X and Y) and computes the
+ *          Root Mean Square Error of the Vector Y with respect to Vector X.
+ *          A boolean (integrated) must also be given to indicate whether the
```

Comment:
Oops, that comment was premature, I know what you mean now. I think that "The boolean integrated" might be clearer, but it is up to you.